C++: Use IR alias analysis for field flow

MathiasVP opened the pull request:
Fixes https://github.com/github/codeql-c-analysis-team/issues/95.

Prior to this PR we wouldn't catch the following flow:

```cpp
struct S {
  int x, y;
};

void test() {
  S s;
  s.x = source();

  S s_copy = s;
  int* px = &s_copy.x;
  sink(*px);
}
```

because the `FieldAddressInstruction` was "far away" from the actual `LoadInstruction`, so we wouldn't recognize that the load generated by `*px` loaded from the field address `x`. This meant that it wasn't picked up as a `readStep` in IR dataflow.

However, the IR already computed that this is a load that totally overlaps with the definition `S s_copy = s`, and so it is reasonably easy to hook into the IR internals and grab this information.

MathiasVP commented:
CPP-difference job started: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1365/

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -145,7 +147,10 @@ OutNode getAnOutNode(DataFlowCall call, ReturnKind kind) {
-  TFieldContent(Field f) or
+  TFieldContent(IntValue startBitOffset, IntValue endBitOffset) {
+    getDefInterval(_, startBitOffset, endBitOffset) or
+    getUseInterval(_, startBitOffset, endBitOffset)
```
Could this be an `and` rather than an `or`? Why track flow through fields that are never written to or never read from?

(I think the shared data-flow library may already do this optimisation, but I'm not sure how late it happens.)

MathiasVP replied:
Good point. I think it's safe to make this a conjunction. I'll try it out locally and check if it makes a performance difference on Jenkins.

MathiasVP replied:
I decided to limit the values in this type differently in 9de570b300dd0cdbe336549d79a4b7b1f1197c99 (instead they're not limited to only those entries that correspond to a `Field`).

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -163,16 +168,18 @@ class Content extends TContent {
-
-  override string toString() { result = f.toString() }
+  override string toString() {
+    result = "[" + startBitOffset.toString() + ".." + endBitOffset.toString() + ")"
```
These strings show up in path explanations, and it would be much less helpful for users if they aren't field names. At least I'm fairly sure they'd show up. The Jenkins job for this PR doesn't show any diff for `ir-path-flow.expected`, but is that just a quirk of how the test are scheduled by `make`?

MathiasVP replied:
You're right (I think they're shown as part of the access path). It should be possible to relate the offsets to a field if we add a third column to `TFieldContent` with a `Type`, and get the old `toString` behavior.

MathiasVP replied:
9de570b300dd0cdbe336549d79a4b7b1f1197c99 now implements a `toString` that should be equal to the version currently on `main`.

jbj reviewed (commented):
I was hoping for a much less disruptive PR where offsets were only used to fix up cases where the field can't be determined by SSA.

MathiasVP commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -2,6 +2,8 @@ private import cpp
 private import semmle.code.cpp.ir.IR
 private import DataFlowDispatch
+private import semmle.code.cpp.ir.implementation.aliased_ssa.internal.AliasedSSA
+private import semmle.code.cpp.ir.internal.IntegerConstant
```
I agree. I'll have a chat with @dbartol about this later today.

MathiasVP commented:
I've added a new iteration of this PR now. Here's an overview:

- 1e13a399323be4a64963d38a143d004d7c6f33eb implements the new API on `ChiInstruction` and `NonPhiMemoryOperand` that @dbartol and I discussed on Zoom the other day.
- 9de570b300dd0cdbe336549d79a4b7b1f1197c99 uses the new API for field flow, and adds a `Class` column to the `FieldContent` branch for more a user-friendly `toString`.
- 3cca74e65455c887d2d77dbf73b0eadd57557d58 accepts the test changes. The new results aren't as impressive as the first iteration of this PR since we added the `Class` column as part of the `FieldContent`, and this means that only source-level fields can be considered `FieldContent`.
This means we don't catch include [lambda-captured variables](https://github.com/github/codeql/blob/main/cpp/ql/test/library-tests/dataflow/dataflow-tests/lambdas.cpp#L46) as fields as we did before adding the `Class` column. In the future we could consider generalizing this column to include "synthesized classes", which could include lambda captures and global variables.

jbj commented:
> This means we don't catch [lambda-captured variables](https://github.com/github/codeql/blob/main/cpp/ql/test/library-tests/dataflow/dataflow-tests/lambdas.cpp#L46) as fields as we did before adding the `Class` column. In the future we could consider generalizing this column to include "synthesized classes", which could include lambda captures and global variables.

Why do we no longer catch lambda-captured variables? I thought they were represented as `Field`s of a synthetic `Class` in our db.

MathiasVP commented:
> Why do we no longer catch lambda-captured variables? I thought they were represented as `Field`s of a synthetic `Class` in our db.

Oh, you're right. They are! I thought that was the reason we caught https://github.com/github/codeql/blob/main/cpp/ql/test/library-tests/dataflow/dataflow-tests/lambdas.cpp#L46 before I added the `Class` column, but not afterwards. I'll do a more thorough investigation of why we lose that new result after the changes in 1e13a399323be4a64963d38a143d004d7c6f33eb and 9de570b300dd0cdbe336549d79a4b7b1f1197c99.

MathiasVP commented:
> Oh, you're right. They are! I thought that was the reason we caught https://github.com/github/codeql/blob/main/cpp/ql/test/library-tests/dataflow/dataflow-tests/lambdas.cpp#L46 before I added the `Class` column, but not afterwards. I'll do a more thorough investigation of why we lose that new result after the changes in [1e13a39](https://github.com/github/codeql/commit/1e13a399323be4a64963d38a143d004d7c6f33eb) and [9de570b](https://github.com/github/codeql/commit/9de570b300dd0cdbe336549d79a4b7b1f1197c99).

Oops! I understand why we're losing that result (along with the others) now. The flow isn't through a lambda-captured variable, but rather through a reference parameter that gets assigned in the body of the lambda. It's the same reason we lose the result in https://github.com/github/codeql/pull/4142/commits/3cca74e65455c887d2d77dbf73b0eadd57557d58#diff-17a3dcef69a5018db3bc4790a32f25b0:

The assignment was recognized as a `storeStep` because `FieldContent` only required the `[begin, end)` offset, so the assignment didn't have to be to a field. But once we introduced the `Class` column in 9de570b300dd0cdbe336549d79a4b7b1f1197c99 we lose that since it's not an assignment to a field.

MathiasVP commented:
I've started a (hopefully final) [CPP-difference](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1384/) that should show that performance is fine now. The internal PR should also be up-to-date.

It would be great if you could comment on 1e13a399323be4a64963d38a143d004d7c6f33eb and 91a23096bb0cd542dac78ecca08d79fd8fddc454 @dbartol.

dbartol commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -144,8 +144,23 @@ OutNode getAnOutNode(DataFlowCall call, ReturnKind kind) {
 predicate jumpStep(Node n1, Node n2) { none() }
 
+/**
+ * Gets the field corresponding to the bit range `[startBit..endBit)` of class `c`.
```
What does this do for the following cases:
```c++
struct SingleElement {
  int x;
};

struct Foo {
  int y;
  SingleElement z;
};

union U {
  int a;
  float b;
};

Foo s;
int w = s.z.x;  // startBit = 32 and endBit = 64, but does that mean "s.z.x" or "s.z"?
U u;
float f = u.b;  // startBit = 0 and endBit = 32, but does that mean "u.a" or "u.b"?
```

MathiasVP replied:
In both cases it gives both results. Specifically, running
```codeql
from Class c, int start, int end, Field f
where f = getField(c, start, end)
select c, start, end, f
```
on your example gives:
```
| test.cpp:5:8:5:10 | Foo | 32 | 64 | test.cpp:2:7:2:7 | x |
| test.cpp:5:8:5:10 | Foo | 32 | 64 | test.cpp:7:17:7:17 | z |
| test.cpp:10:7:10:7 | U | 0 | 32 | test.cpp:11:7:11:7 | a |
| test.cpp:10:7:10:7 | U | 0 | 32 | test.cpp:12:9:12:9 | b |
```

jbj replied:
We don't want multiple results from `FieldContent.toString`. You could use `min` or `strictconcat` to collapse them down to one.

MathiasVP replied:
Fixed in 01bf051.

dbartol commented on cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll:
```diff
@@ -1962,6 +1962,13 @@ class ChiInstruction extends Instruction {
   final Instruction getPartial() { result = getPartialOperand().getDef() }
+
+  /**
+   * Gets the bit range `[startBit, endBit)` updated by the partial operand of this `ChiInstruction`.
```
```suggestion
   * Gets the bit range `[startBit, endBit)` updated by the partial operand of this `ChiInstruction`, relative to the start address of the total operand.
```

dbartol reviewed (approved):
I think this PR is ready to go once tests are passing. However, after talking to @rdmarsh2, I'm not sure that exposing the bit offsets is the right level of abstraction. I've opened https://github.com/github/codeql-c-analysis-team/issues/138 for possible follow-up.

Let's merge this PR now, though, because it does the right analysis and gets pretty much the right results. We can worry about cleaning up the API and alert messages separately.

MathiasVP commented:
> I think this PR is ready to go once tests are passing. However, after talking to @rdmarsh2, I'm not sure that exposing the bit offsets is the right level of abstraction. I've opened [github/codeql-c-analysis-team#138](https://github.com/github/codeql-c-analysis-team/issues/138) for possible follow-up.

I think the only failing test now is the change in the internal tests.

MathiasVP commented:
> I've started a (hopefully final) [CPP-difference](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1384/) that should show that performance is fine now. The internal PR should also be up-to-date.

The CPP-difference shows a 30% performance regression on Kamailo, and a 10% regression on Linux. I'll investigate at these later today.

dbartol commented:
Is this really supposed to be marked as "Depends on internal PR"?

MathiasVP commented:
> Is this really supposed to be marked as "Depends on internal PR"?

Yes. Once we merge it we should also merge the internal PR https://github.com/github/semmle-code/pull/37593 that contains the changes to the internal test.

MathiasVP commented:
Hopefully, 01bf051 fixes the last performance problems. Here's a (final?) CPP-difference run: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1388/

MathiasVP commented:
Oops. I was too trigger happy with my rebase from `main`. Sorry!

MathiasVP commented:
> Sorry for the annoyingly long git conversation this has caused 😢. Anyway, assuming the [CPP-differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1392/) comes back without performance regressions this should be good to go.

CPP-difference came back with good performance.

MathiasVP commented:
> Sorry for the annoyingly long git conversation this has caused 😢. Anyway, assuming the [CPP-differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1392/) comes back without performance regressions this should be good to go.

The CPP-difference showed no performance regressions! 🎉

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -163,17 +178,18 @@ class Content extends TContent {
-  override predicate hasLocationInfo(string path, int sl, int sc, int el, int ec) {
-    f.getLocation().hasLocationInfo(path, sl, sc, el, ec)
-  }
+  Field getField() { result = getField(c, startBit, endBit) }
```
I suggest renaming these two `getField` predicates to `getAField` so it's clear to callers there might be multiple results.

MathiasVP replied:
Good point. I've done this in 9c1837e3b31b7bbc017c7e8d583de08fc88270e3. I also changed the qldoc to reflect that there might be multiple results.

jbj commented on cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp:
```diff
@@ -468,7 +468,7 @@ void test_swop() {
 
 	sink(x); // clean [FALSE POSITIVE]
-	sink(y); // tainted
+	sink(y); // tainted [NOT DETECTED by IR]
```
Do we have an internal issue for recovering these cases of lost flow?

jbj reviewed (commented):
Otherwise LGTM.

jbj reviewed (dismissed):
LGTM. I'll leave the final review/merge to @dbartol.

dbartol reviewed (dismissed):
The cleaned-up IR changes look good. LGTM.

jbj commented:
There's a conflict in a test output file.

MathiasVP commented:
Note that the only check that's not passing is one that's broken on all PRs. (So this PR should be safe to merge.)
